core/00_primordials.js

```diff
@@ -479,27 +480,76 @@
   primordials.PromisePrototypeCatch = (thisPromise, onRejected) =>
     PromisePrototypeThen(thisPromise, undefined, onRejected);
 
+  const arrayToSafePromiseIterable = (array) =>
+    new SafeArrayIterator(
+      ArrayPrototypeMap(
+        array,
+        (p) => {
+          if (ObjectPrototypeIsPrototypeOf(PromisePrototype, p)) {
+            return new SafePromise((c, d) => PromisePrototypeThen(p, c, d));
+          }
+          return p;
+        },
+      ),
+    );
+
   /**
    * Creates a Promise that is resolved with an array of results when all of the
    * provided Promises resolve, or rejected when any Promise is rejected.
-   * @param {unknown[]} values An array of Promises.
-   * @returns A new Promise.
+   * @template T
+   * @param {Array<T | PromiseLike<T>>} values
+   * @returns {Promise<Awaited<T>[]>}
    */
   primordials.SafePromiseAll = (values) =>
     // Wrapping on a new Promise is necessary to not expose the SafePromise
     // prototype to user-land.
     new Promise((a, b) =>
-      SafePromise.all(
-        ArrayPrototypeMap(
-          values,
-          (p) => {
-            if (ObjectPrototypeIsPrototypeOf(PromisePrototype, p)) {
-              return new SafePromise((c, d) => PromisePrototypeThen(p, c, d));
-            }
-            return p;
-          },
-        ),
-      ).then(a, b)
+      SafePromise.all(arrayToSafePromiseIterable(values)).then(a, b)
+    );
+
+  /**
+   * Creates a Promise that is resolved with an array of results when all
+   * of the provided Promises resolve or reject.
+   * @template T
+   * @param {Array<T | PromiseLike<T>>} values
+   * @returns {Promise<PromiseSettledResult<T>[]>}
+   */
+  primordials.SafePromiseAllSettled = (values) =>
```

Comment:
I assume these 3 APIs are added without actual usages at this point. I think we should add these APIs only when they are actually necessary in implementation

Reply:
I am concerned that when they are needed, `SafePromiseAllSettled` etc. will not be introduced and an insecure `PromiseAllSettled` etc. will be used.

Reply:
@kt3k I suggest leaving it in the code as a comment out so that it can be exported when needed.

Reply:
> @kt3k I suggest leaving it in the code as a comment out so that it can be exported when needed.

That makes sense to me.